### ISSUE TITLE
Training my pretrained classification model on custom dataset with fresh dataset while retaining the weights from previous learnings

### ISSUE THREAD

**Issue by coder-glitche** [ultralytics/ultralytics#4005]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello @glenn-jocher i was recently trying to retrain my yolov8 classification model which was already trained on my custom dataset now i want to train the same model with same classes but fresh dataset what should i do please help me im starting out and exploring any resources for yolov8 transfer learning would be helpful

**Comment by glenn-jocher** [comment]
@coder-glitche hello, and thanks for your question!

You are on the right track starting with transfer learning. When you have a pretrained model, and you want to keep training that model on a new dataset while retaining the previously learned weights, what you're doing is precisely transfer learning.

YOLOv8 supports transfer learning. You just need to use your pretrained model as the --weights argument during the training process. The rest of the training process will proceed the same way as a normal training process.

To summarize, you'll need to do the following steps:

1. Prepare your fresh dataset just like you would for a normal training process.
2. Begin training, but this time pass to the --weights argument the path where your pretrained model is located.
3. YOLOv8 will initialize the model with the weights from your pretrained model, and continue the training process from there.

This way, your model will keep learning from the new data, while still retaining the knowledge it acquired from the previous training sessions. 

We hope this information helps you on your journey. Good luck, and feel free to ask any further questions!

**Comment by SujithChristopher** [comment]
Hi @glenn-jocher , 

I'm currently not using CLI, and i could not find the "weights" argument in the documentation, but there is "load" argument, are those the same?

`model = YOLO('yolov8n-pose.pt').load("pretrained.pt")`
`model.train(data='dataset.yaml', epochs=100)`

is this the correct format for transfer learning with previously trained model, (with the same classes).

**Comment by glenn-jocher** [comment]
Hi @SujithChristopher,

Yes, you are correct! In the context of the Python usage, we use the `.load()` function to load our pretrained model's weights. Your provided code snippet is indeed a representation of how to apply transfer learning using a previously trained model with the Ultralytics YOLOv8 API in Python:

```python
model = YOLO('yolov8n-pose.pt').load("pretrained.pt")
model.train(data='dataset.yaml', epochs=100)
```

Here, the `YOLO('yolov8n-pose.pt')` function initializes the YOLO model architecture, then the `.load("pretrained.pt")` function loads the weights from your pretrained model. The `.train(data='dataset.yaml', epochs=100)` function then kicks off the training process on your new dataset. 

So, you've got it right! Let me know if you have any other questions, we're here to help!

**Comment by SujithChristopher** [comment]
Thank you @glenn-jocher for confirmation.

**Comment by glenn-jocher** [comment]
You're welcome, @SujithChristopher! I'm glad I could help. Don't hesitate to reach out if you have more questions or if there's anything else you're curious about. Happy coding with YOLOv8!

**Comment by yash-seth** [comment]
Hi @glenn-jocher , I was curious to know if there was a way I could keep the pre-trained weights and the 80 classes of COCO that the model has been trained on + say 10 classes of my own from a custom dataset? So in totality, the model should be able to detect 80 + 10 = 90 classes and have relevant weights for all 90 classes? Thanks in advance for your help.

**Comment by glenn-jocher** [comment]
@yash-seth hello! To retrain your YOLOv8 classification model on a new dataset while retaining the weights from previous learnings, you can follow these steps:

1. **Prepare Your New Dataset**: Ensure your new dataset is formatted correctly for YOLOv8. It should have the same class structure as the previous dataset to maintain class consistency.

2. **Load Your Pretrained Model**: Load your pretrained model using the `YOLO` class. You can specify the path to your model's `.pt` file.

3. **Train the Model**: Use the `train` method to start training. You can specify your new dataset configuration file and any other training parameters you wish to adjust.

Here's an example of how you can do this in Python:

```python
from ultralytics import YOLO

# Load your pretrained classification model
model = YOLO('path/to/your/pretrained_model.pt')

# Train the model on your new dataset
# Make sure to specify the path to your new dataset configuration file
results = model.train(data='path/to/your/new_dataset.yaml', epochs=5)

# Optionally, you can specify additional training arguments such as batch size, image size, etc.
```

If you're using the CLI, the command would look something like this:

```bash
yolo classify train model=path/to/your/pretrained_model.pt data=path/to/your/new_dataset.yaml epochs=5
```

Remember to replace `path/to/your/pretrained_model.pt` with the actual path to your pretrained model file and `path/to/your/new_dataset.yaml` with the path to your new dataset configuration file.

### Transfer Learning Tips:

- **Maintain Class Structure**: Ensure the class labels in your new dataset match those in the original dataset to avoid confusion during training.
- **Learning Rate**: Consider using a lower learning rate since you're fine-tuning a pretrained model.
- **Freezing Layers**: Depending on your dataset size and similarity to the original dataset, you might want to freeze some of the earlier layers during training to retain the learned features.

By following these steps, you should be able to retrain your YOLOv8 classification model on a new dataset while leveraging the knowledge it has gained from the previous training sessions.

**Comment by yash-seth** [comment]
Hi @glenn-jocher , I wished to understand what happens to the output layer when one trains a pretrained model on a custom dataset. Does the output layer get removed and a new output layer is added as per the custom dataset?

Since, say COCO has 80 or so classes, so then the pretrained model would have 80 output nodes, while if my custom dataset has 4 classes, then the output layer is required to have 4 output nodes only.

The above query is with respect to what happens to the output layer when this command is run on a custom dataset,

`!yolo task=detect mode=train data=data.yaml model=yolov8s.pt epochs=20 lr0=0.01`

**Comment by glenn-jocher** [comment]
@yash-seth hi there! 👋 Great question about how the output layer adapts during training on a custom dataset with a different number of classes.

When you train a pretrained model (like one trained on COCO with 80 classes) on a custom dataset with a different number of classes (say 4 in your case), YOLOv8 automatically adjusts the output layer to match the number of classes in your custom dataset. This adjustment happens seamlessly when you specify your dataset in the `data.yaml` file, which should include the number of classes for your custom dataset.

Here's a quick overview of what happens:
- The pretrained model's final layer, designed for the original number of classes (80 for COCO), is replaced with a new layer that has the correct number of output nodes for your dataset (4 in your example).
- This new output layer is then trained along with the rest of the model on your custom dataset.

So, when you run the command:

```bash
!yolo task=detect mode=train data=data.yaml model=yolov8s.pt epochs=20 lr0=0.01
```

The model's architecture is automatically adjusted to accommodate the number of classes specified in your `data.yaml`. Just make sure your `data.yaml` correctly specifies the number of classes for your custom dataset. Happy training! 🚀

**Comment by glenn-jocher** [comment]
@yash-seth you're welcome! If you have any more questions or need further clarification, feel free to ask. Happy coding! 😊

**Comment by Neilam052119** [comment]
hi @glenn-jocher  i am trying to train new dataset using pretrained model weights (YOLOv5 model) , but it could not run path of  DATASET , though it  is present in my laptop . DATASET is folder of images converted from pdf to images from various financial  newspaper across world .  basically data to pretrain is images and i need to fine tune to get new weights .pt file

**Comment by glenn-jocher** [comment]
Hello! It sounds like you're encountering a path issue with your dataset. Here are a few things you might want to check:

1. **Path Accuracy**: Ensure the path you've specified in your dataset configuration file exactly matches the location of your dataset on your laptop.
2. **Dataset Format**: Verify that your dataset is structured correctly according to the YOLOv5 requirements. Each class should have its own folder containing all the images for that class.
3. **Permissions**: Make sure you have the necessary permissions to access the folders and files in your dataset.

If everything seems correct and you're still facing issues, could you please share the error message you're receiving? This will help in diagnosing the problem more effectively.
